rs/sns/treasury_manager/src/lib.rs

```diff
@@ -0,0 +1,286 @@
+use candid::{CandidType, Nat, Principal};
+use serde::{Deserialize, Serialize, Serializer};
+use std::{collections::BTreeMap, fmt::Display};
+
+#[derive(CandidType, Clone, Debug, Deserialize)]
+pub struct TreasuryManagerInit {
+    pub allowances: Vec<Allowance>,
+}
+
+#[derive(CandidType, Clone, Debug, Deserialize)]
+pub struct TreasuryManagerUpgrade {}
+
+#[derive(CandidType, Clone, Debug, Deserialize)]
+pub enum TreasuryManagerArg {
+    Init(TreasuryManagerInit),
+    Upgrade(TreasuryManagerUpgrade),
+}
+
+#[derive(CandidType, Clone, Debug, Deserialize, PartialEq)]
+pub struct Balance {
+    #[serde(serialize_with = "serialize_nat_as_u64")]
+    pub amount_decimals: Nat,
+    pub account: Option<Account>,
+}
+
+#[derive(CandidType, Clone, Debug, Deserialize, Eq, Ord, PartialEq, PartialOrd)]
+pub enum Party {
+    TreasuryOwner,
+    TreasuryManager,
+    External,
+    LedgerFee,
+}
+
+#[derive(CandidType, Clone, Debug, Deserialize, PartialEq)]
+pub struct BalancesForAsset {
+    pub party_to_balance: Option<BTreeMap<Party, Balance>>,
+}
+
+#[derive(CandidType, Clone, Debug, Default, Deserialize, PartialEq)]
+pub struct Balances {
+    pub timestamp_ns: u64,
+    pub asset_to_balances: Option<BTreeMap<Asset, BalancesForAsset>>,
+}
+
+pub type TreasuryManagerResult = Result<Balances, Vec<TransactionError>>;
+
+pub trait TreasuryManager {
+    /// Implements the `deposit` API function.
+    fn deposit(
+        &mut self,
+        request: DepositRequest,
+    ) -> impl std::future::Future<Output = TreasuryManagerResult> + Send;
+
+    /// Implements the `withdraw` API function.
+    fn withdraw(
+        &mut self,
+        request: WithdrawRequest,
+    ) -> impl std::future::Future<Output = TreasuryManagerResult> + Send;
+
+    /// Implements the `audit_trail` API query function.
+    fn audit_trail(&self, request: AuditTrailRequest) -> AuditTrail;
+
+    /// Implements the `balances` API query function.
+    fn balances(&self, request: BalancesRequest) -> TreasuryManagerResult;
+
+    /// Should not be exposed as an API function, but rather called periodically by the canister.
+    fn refresh_balances(&mut self) -> impl std::future::Future<Output = ()> + Send;
+
+    /// Should not be exposed as an API function, but rather called periodically by the canister.
```

Comment:
same question here: what does it do? redistribute the rewards from the LP? But where to - jsut back to the main treasury?

Reply:
Similarly, this is to guide developers in which periodic tasks are required for a treasury manager implementation.